torchrl/collectors/collectors.py

```diff
@@ -124,6 +128,28 @@ def _get_policy_and_device(
                     "env must be provided to _get_policy_and_device if policy is None"
                 )
             policy = RandomPolicy(self.env.action_spec)
+        elif (
+            observation_spec is not None
+            and isinstance(policy, nn.Module)
+            and not isinstance(policy, TensorDictModule)
+        ):
+            # auto-wrap policy with TensorDictModule
+            sig = inspect.signature(policy.forward)
+            next_observation = {
+                key[5:]: value
+                for key, value in observation_spec.rand().items()
+                if key.startswith("next_")
+            }
+            if set(sig.parameters) == set(next_observation):
+                out_keys = ["action"]
+                output = policy(**next_observation)
+
+                if isinstance(output, tuple):
+                    out_keys.extend(f"output{i+1}" for i in range(len(output) - 1))
+
+                policy = TensorDictModule(
+                    policy, in_keys=list(sig.parameters), out_keys=out_keys
+                )
```

Comment:
shouldn't we raise and exception if a policy is not compatible with the env specs?

Reply:
Sure, makes sense, so user must either pass:

- a `TensorDictModule`
- a `nn.Module` that can read `TensorDicts` as in your comment above
- a `nn.Module` that is compatible with the env and can be auto-wrapped

And if none of those conditions are met, we can reasonably raise a `TypeError`? There's no other case we need to support?

Reply:
Added in some exceptions + tests that they get raised when appropriate.